src/Expect.elm

```diff
@@ -246,6 +255,84 @@ atLeast =
     compareWith "Expect.atLeast" (>=)
 
 
+{-| A type to describe how close a floating point number must be to the expected value for the test to pass. This may be
+specified as absolute or relative. If you're not deeply familiar with binary floating point, you probably want `Absolute`.
+
+`AbsoluteOrRelative` tolerance uses a logical OR between the absolute (specified first) and relative tolerance. If you
+want a logical AND, use [`Expect.all`](#all).
+
+-}
+type FloatingPointTolerance
+    = Absolute Float
+    | Relative Float
+    | AbsoluteOrRelative Float Float
+
+
+{-| Passes if the second and third arguments are equal within a relative tolerance
+specified by the first argument. This is intended to avoid failing because of
+minor inaccuracies introduced by floating point arithmetic.
+
+    -- Fails because 0.1 + 0.2 == 0.30000000000000004 (0.1 is non-terminating in base 2)
+    0.1 + 0.2 |> Expect.equal 0.3
+
+    -- So instead write this test, which passes
+    0.1 + 0.2 |> Expect.within 0.000000001 0.3
+
+Failures resemble code written in pipeline style, so you can tell
+which argument is which:
+
+    -- Fails because 3.14 is not close enough to pi
+    3.14 |> Expect.within 0.0001 pi
+
+    {-
+
+    3.14
+    ╷
+    │ Expect.within 0.0001
+    ╵
+    3.141592653589793
+
+    -}
+
+-}
+within : FloatingPointTolerance -> Float -> Float -> Expectation
+within tolerance =
+    compareWith ("Expect.within " ++ toString tolerance)
+        (withinCompare tolerance)
+
+
+{-| Passes if (and only if) a call to `within` with the same arguments would have failed.
+-}
+notWithin : FloatingPointTolerance -> Float -> Float -> Expectation
+notWithin tolerance =
+    compareWith ("Expect.notWithin " ++ toString tolerance)
+        (\a b -> not <| withinCompare tolerance a b)
+
+
+withinCompare : FloatingPointTolerance -> Float -> Float -> Bool
+withinCompare tolerance a b =
+    let
+        ( absoluteTolerance, relativeTolerance ) =
+            case tolerance of
+                Absolute absolute ->
+                    ( max 0 absolute, 0 )
+
+                AbsoluteOrRelative absolute relative ->
+                    ( max 0 absolute, max 0 relative )
+
+                Relative relative ->
+                    ( 0, max 0 relative )
```

Comment:
It seems like we are throwing away an interesting error condition, a nonpositive tolerance. Maybe we should catch that early and provide a useful failure message?